specs/eid-hermes-commands-format.md

```diff
@@ -0,0 +1,227 @@
+# eid-hermes Commands
+
+## Introduction
+
+This document defines the commands that can be sent to a Hermes device. It also
+defines the command completions which will be sent by the device back to the
+host. It includes both the commands' format as well as which commands are
+available.
+
+## Sample Execution
+
+The following diagram exemplifies a possible execution flow for Hermes. Note
+how the data slot is reused to process data in chunks at a time.
+
+```
+┌--------------┐     ┌---------------┐    ┌--------------┐     ┌------------┐
+| Request slot | --> | Write program | -->| Request slot | --> | Write data |<-┐
+|   (program)  |     └---------------┘    |    (data)    |     └------------┘  |
+└--------------┘                          └--------------┘             |       |
+                                                                       v       |
+┌--------------┐     ┌--------------┐     ┌--------------┐     ┌-------------┐ |
+| Release slot | <-- | Release slot | <-- |   Read data  | <-- | Run program | |
+|  (program)   |     |    (data)    |     |     slot     |     └-------------┘ |
+└--------------┘     └--------------┘     └--------------┘                     |
+                                                  |                            |
+                                                  └----------------------------┘
+```
+
+## Command Format
+
+Command Requests are 32 bytes long and follow the format of Table 1. Command
+Responses are 16 bytes long and follow the format of Table 2. Some parts of
+Requests and Responses are command-specific and are described below.
+
+| Bytes | Description                              |
+|-------|------------------------------------------|
+| 00    | Opcode: defines which command to execute |
+| 01    | Reserved                                 |
+| 03:02 | Command Identifier: a Hermes user should never repeat this field. Different users may repeat an ID (the driver has other ways to differentiate between users). It is undefined behaviour to reuse an ID. |
```

Comment:
This is only 255 IDs so I assume at some point they have to be reused?

Reply:
This is 2 bytes, so 65535 IDs (the NVMe spec also uses 2 bytes for CID). Not sure how it handle ID reuse...

Reply:
If we're using a sync model without queues, do we need to keep the command identifier?

Reply:
@iomartin I suggest we keep the command identifier and just use a rolling counter in the driver.

Reply:
@iomartin I would say "Command Identifier: The hermes device will copy this value into the corresponding Command Response." and leave it at that.